libs/runtime/ebpf_epoch.c

```diff
@@ -4,130 +4,174 @@
+ */
 static uint32_t _ebpf_epoch_cpu_count = 0;
 
 /**
- * @brief _ebpf_current_epoch indicates the newest active epoch. All memory free
- * operations were performed prior to this value.
+ * @brief Enum of messages sent between CPUs.
+ */
+typedef enum _ebpf_epoch_cpu_message_type
+{
+    EBPF_EPOCH_CPU_MESSAGE_TYPE_PROPOSE_RELEASE_EPOCH, ///< This message is sent to CPU 0 to propose a new release
+                                                       ///< epoch.
+                                                       ///< CPU 0 declares the new current epoch and proposes it as the
+                                                       ///< release epoch. Each CPU then queries the epoch for each
+                                                       ///< thread linked to this CPU and sets the proposed release
+                                                       ///< epoch in the message to the minimum of the local minima and
+                                                       ///< the minima in the message. The message is then forwarded to
+                                                       ///< the next CPU. The last CPU then sends an epoch commit
+                                                       ///< message to CPU 0 with the final proposed release epoch.
+
+    EBPF_EPOCH_CPU_MESSAGE_TYPE_COMMIT_RELEASE_EPOCH, ///< This message is sent to CPU 0 to commit the proposed release
+                                                      ///< epoch.
+                                                      ///< Each CPU then:
+                                                      ///< 1. Clears the timer-armed flag.
+                                                      ///< 2. Sets the released epoch to the proposed release epoch
+                                                      ///< minus 1.
+                                                      ///< 3. Releases any items in the free list that are eligible for
+                                                      ///< reclamation.
+                                                      ///< 4. Rearms the timer if need.
+                                                      ///< 5. Forwards the message to the next CPU.
+                                                      ///< The last CPU then sends a epoch computation complete message
+                                                      ///< to CPU 0.
+    EBPF_EPOCH_CPU_MESSAGE_TYPE_PROPOSE_EPOCH_COMPLETE, ///< This message is sent only to CPU 0 to signal that epoch
+                                                        ///< computation is complete.
+    EBPF_EPOCH_CPU_MESSAGE_TYPE_EXIT_EPOCH, ///< This message is used when a thread running with IRQL < DISPATCH calls
+                                            ///< ebpf_epoch_exit on a different CPU than ebpf_epoch_enter. It is sent
+                                            ///< during from the thread that is existing the epoch to the CPU that
```

Comment:
Can't parse "during from".
Also "existing" should probably be "exiting"

Reply:
Hopefully it's clearer now.